### DISCUSSION TITLE
[API Proposal:] scoped out Span<T>

### DISCUSSION THREAD

**Discussion by CodingMadness** [dotnet/csharplang#6669] [Language Ideas]
Heyo,

would a construct like this be possible?

```
public void DoSomeSpanStuff(scoped out Span<char> begin)
{
     begin = stackalloc char[512*2];  
    //OR
    begin = stackalloc char[ComputeSize()];
}

OuterMethod()
{
      DoSomeSpanStuff(out var span);
}
```

and internally, rosyln could create something like this, 

```
OuterMethod() 
{
    int _size = Compute_Size_function(); //if its dynamically created, the respecitve method inside the DOSpanStuff() could be inlined here and if its not just inline the expression directly here
    begin = stackalloc char[_size];
    DoSpanStuff(begin);
}
```

Any discussion is wished!

Cheers

**Comment by svick** [comment]
I agree this is definitely something worth looking into (see https://github.com/dotnet/csharplang/discussions/1817 for my own wild ideas on the topic), but I'm not sure this exact approach would work, because it would be hard to extract the `Compute_Size_function()` from the code.

Consider even something simple like the following:

```c#
void ReadStream(Stream stream, scoped out Span<byte> result)
{
    int length = (int)stream.Length;

    if (length <= 128)
    {
        result = stackalloc byte[length];
    }
    else
    {
        result = new byte[length];
    }

    stream.ReadExactly(result);
}
```

How would `Compute_Size_function()` look like in this case? And what if the `length` had to be read from the `stream`, how would it look like then?

What might work is limiting the size of the `stackalloc` to some constant. For example:

```c#
void ReadStream(Stream stream, [AutoStackallocSize(128)] scoped ref Span<byte> result)
{
    int length = (int)stream.Length;

    if (length <= 128)
    {
        result = result.Slice(length);
    }
    else
    {
        result = new byte[length];
    }

    stream.ReadExactly(result);
}
```

Calling this function like this:

```c#
ReadStream(stram, out var result);
```

would compile to:

```c#
Span<byte> result = stackalloc byte[128];
ReadStream(stream, ref result);
```

And since stack allocating a constant amount is the recommended practice, this shouldn't even be that limiting.

I'm not sure how exactly to reconcile the fact that the auto-stack allocated parameter would look like `ref` from the inside, but like `out` from the outside, but I don't think that's a significant hurdle.

**Reply by FaustVX** [reply]
I like the `AutoStackallocSizeAttribute` approach, but what if `ReadStream` return the array instead, The `stackalloc` will still be allocated on the stack but won't be used.

**Reply by svick** [reply]
@FaustVX Yes, but I don't see a good way around that. I also don't think that it's a big deal, since `stackalloc`s are quite cheap.

**Reply by CodingMadness** [reply]
@svick 

```
static int count = 2;

static void DoSomething(){
  //...... other stuff......
  count++;
}

static int GetSize() => count;

ref struct SomeStruct{
  SomeStruct()
  {
    // GetSize() => 2;
    DoSomething(); 
    // GetSize() => 3;
    stackalloc[GetSize()];
    WorkWithStackAllocation();
  }

  

  SomeStructPart1()
  {
    // GetSize() => 2;
    DoSomething();
    //Store Current Method State 
  }

  SomeStructPart2()
  {
    //Load current method state
    WorkWithStackAllocation();
  }

  void WorkWithStackAllocation(){
    //Important only that the span is safe to escape here!
  }
}

void CreateStruct()
{
  var instance = new SomeStruct();

  //generated alternative
  var instance = default;
  instance.SomeStructPart1();
  instance.Span = stackalloc[GetSize()];
  instance.SomeStructPart2();
}
```

I think this would be a decent opportunity, to allow **a)** variable size length to a certain minimum and **b)** ofc const size. The only thing this approach does not handle is branch-prediction, i thought that would be out of scope for this. but 2/3 usecases are handled.

**Reply by CodingMadness** [reply]
This allows you to have stackalloc of T inside a ref-struct. maybe similar stuff could be done in terms of Code-Gen for the "scoped out Span<T>" pattern.

**Reply by svick** [reply]
@Shpendicus It sounds to me like you're describing effectively turning the function into a coroutine. That would work (it's Approach 1 in my https://github.com/dotnet/csharplang/discussions/1817), but I now think it's way too complicated for this.

**Reply by CodingMadness** [reply]
> @Shpendicus It sounds to me like you're describing effectively turning the function into a coroutine. That would work (it's Approach 1 in my #1817), but I now think it's way too complicated for this.

possible mate, but for now i dont have other idea , this is such a neat feature to have. Because as it is now, you have to either forceably take a managed array or let the user define ur stackalloc which is nonesense.

**Comment by AraHaan** [comment]
I feel like this is not good (at it's current state), as effectively what if it's under allocating anything, also consider this, what if the GC decides to move the ref around too and you try to use the ref (not in a stream or w/e) in a native p/invoke call? In the case of p/invoke calls it could be disastrous when the code simply blows up right when it tries to read memory but while it does the GC decided to move it (because how will the native code know that it did and instead read it from the updated location)? So, while this looks nice for managed code. I feel that it should be limited to managed stackallocs and that computing sizes should be done by the member that allocates (as usually they tend to allocate exactly what they need).

**Reply by Joe4evr** [reply]
> (as usually they tend to allocate exactly what they need).

Agreed on all counts except this last bit can do with an addendum: You always want to use a compile-time constant for how much you want to stackalloc. If you were creating a variable-size stack buffer, then even if you properly mitigate the odds of a stack overflow, it's way more expensive for the JIT to produce than a const-initialized buffer.

**Reply by CodingMadness** [reply]
> I feel like this is not good (at it's current state), as effectively what if it's under allocating anything, also consider this, what if the GC decides to move the ref around too and you try to use the ref (not in a stream or w/e) in a native p/invoke call? In the case of p/invoke calls it could be disastrous when the code simply blows up right when it tries to read memory but while it does the GC decided to move it (because how will the native code know that it did and instead read it from the updated location)? So, while this looks nice for managed code. I feel that it should be limited to managed stackallocs and that computing sizes should be done by the member that allocates (as usually they tend to allocate exactly what they need).

Aren't refs especially made for such purposes? To be updated once the GC kicks in? and whenever that may be, for your p/invoke you still do smth like "myRef.AsPointer()" and the moment you do that, you will end up always with a valid address. 

And also I think that you can "pin" the adress while you pass it to the native call, so whatever there happens with it, like reading chunks from it, the pointer/ref can be pinned for that duration.

**Reply by CodingMadness** [reply]
BTW: i just realize, why are we talking even about GC ? this is a proposal to allow stackallocated bufffer to the outer stackframe, this is unmanaged territory, there aint any GC happening.

**Comment by CodingMadness** [comment]
Does anyone from the .NET team has taken atleast a slight interesst in this and if not, why?

**Reply by CyrusNajmabadi** [reply]
I imagine if someone takes interest, they will comment here.  :)

**Reply by CodingMadness** [reply]
apparently there is not enough interesst then for it which i find to be sad, cause it can enhance the usage of the stack more in certain scenarios.

**Reply by CyrusNajmabadi** [reply]
I'd work on beefing up the proposal.  including figuring out how it would work.  A prototype might also be a good idea, demonstrating that it's possible and which shows the advantages.

**Reply by CodingMadness** [reply]
@CyrusNajmabadi thx for the response so far mate, hmm okay will make a bigger investment and bring up more details in :--)

**Reply by RikkiGibson** [reply]
Note that the particular terminology of `scoped out T param` may cause confusion. This would tend to mean "you can't return `param` by reference". Not "the callee stack-allocates something with a longer lifetime than the call".
